### ISSUE TITLE
Create scripts from a template like Godot

### ISSUE THREAD

**Issue by DrakeFruit** [Facepunch/sbox-scenestaging#8]
When making a new script in Godot (would be a component in S&box) you have the option to choose which programming language it's in, the folder, name, and what template it's based off if any in a neat little GUI. Not sure what the best way to implement this would be but being able to select a template from a dropdown when creating a new component would be really nice.  Prevents cluttering the components menu and removes and potential copy and pasting or overriding bullshit.
https://github.com/Facepunch/sbox-scenestaging/assets/94070205/9e299817-d7dd-4c7d-926d-2a2f37134b4f

**Comment by CarsonKompon** [comment]
Now that we have the following:
![image](https://github.com/Facepunch/sbox-scenestaging/assets/5159369/1332b58e-d88b-4304-a0d2-e0687f36a9a9)

It would be nice if those options could be expanded to:
"New {x} Component from {y}..."
(with y being the template)

So you could have:
![image](https://github.com/Facepunch/sbox-scenestaging/assets/5159369/3dfd49e8-e8eb-4ea6-9f13-aa5f9f28c8dd)

**Comment by garrynewman** [comment]
I think we should do this.. right now it's hard to explain to people how to create UI components. This would solve that.

![image](https://github.com/Facepunch/sbox-scenestaging/assets/928843/cd9193b1-ca0e-4fb9-9d38-3ef23f4b9e96)

Process would be:

- write the name of the component (auto focused text entry)
- a file path will be auto filled in
- optionally change the path (folder popup to choose). 
- optionally change the component type template
- 
For the UI I'm picturing something like the old Visual Studio - New Project dialog, but simpler.

![image](https://github.com/Facepunch/sbox-scenestaging/assets/928843/ee9df14b-91de-4ebc-a82b-0b6bc3369efc)

# Templates

### Simple Component

{filename}.cs
```
using Sandbox;

public sealed class {{componentName}} : BaseComponent
{
	public override void Update()
	{
		
	}
}
```


### Razor Panel Component

{filename}.razor
```
@using Sandbox;
@using Sandbox.UI;
@inherits PanelComponent

<root>
    <div class="title">@MyStringValue</div>
</root>

@code{

    [Property] public string MyStringValue { get; set; } = "Hello World!";

    /// <summary>
    /// the hash determines if the system should be rebuilt. If it changes, it will be rebuilt
    /// </summary>
    protected override int BuildHash() => System.HashCode.Combine( MyProperty );
}
```

{filename}.razor.scss
```

{{componentName}}
{
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: #444;
    justify-content: center;
    align-items: center;
    font-weight: bold;
    border-radius: 20px;

    .title
    {
        font-size: 25px;
        font-family: Poppins;
        color: #fff;
    }
}
```

**Comment by DevulTj** [comment]
I've added that now. I went with a simpler UI to begin with, it's currently in-line with the Compoment dialog. 

![image](https://github.com/Facepunch/sbox-scenestaging/assets/22372616/feab5d80-984b-4c70-bc88-7a3afc3b58ad)

Let me know if this UI is not showing enough info, and I'll see what I can do.

Developers can define new component templates by inheriting from the `ComponentTemplate` class. (Not sure if we want to go down the same route as Project Templates for this with loose files, might be too much)
